### PR TITLE
Add Homogeneous and HomogeneousRowList for homogeneous rows

### PR DESCRIPTION
**Pull request by paulyoung**
This is my first attempt at type-level programming. Hopefully it's pretty self explanatory, but the idea is to make sure that every field in a record has the same type.

Something similar can be achieved using `StrMap`, but it's not quite the same.

This initial version is available to play with at http://try.purescript.org/?gist=bec330aab7a6c0b21e466e37483dfba9. It was originally adapted from http://try.purescript.org/?gist=2720825b0476c8924717437fb6f6eefb.

I think it would be great to support polymorphism and open rows, but I haven't had any success with that.

Any and all feedback is welcome.

### CODE REVIEW

File: src/Type/Data/Record.purs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  ) where
+
+import Type.Row (class RowToList, Cons, Nil, kind RowList)
+
+class RecordOf record fieldType
+instance recordOf
+  :: ( RowToList record fields
+     , FieldOf fields fieldType
+     )
+  => RecordOf (Record record) fieldType
```
Comment (paulyoung) [inline src/Type/Data/Record.purs]: I think `(Record record)` is the right thing to do here (as opposed to just `record`) but it has the consequence of changing the type of a function beyond simply adding the type constraint.

For example, this:

```purescript
foo :: forall a
     . Record a
    -> Record a
```

becomes:

```purescript
foo :: forall a
     . RecordOf a Boolean
    => a
    -> a
```

**Comment by paf31** [comment]
Looks good, thanks! A few notes though:

- I'd rather call this something like `Homogeneous`, and have it be a predicate on the row type, not the record type.
- It should use `TypeEquals`, to force the types to be equal by unification.
- I think you can add a fundep from the row to the type in the labels, which might be useful.

**Comment by LiamGoodacre** [comment]
@paf31 is the motivation for the `TypeEquals` approach, that you get a nicer error and don't see the `RowList` implementation artifact?

**Comment by paf31** [comment]
Yes, since the type is already determined if the fundep is added anyway.

**Comment by paulyoung** [comment]
Thanks for the feedback so far. Good call on moving this to `Row`.

Here's what I have so far:

```purescript
class Homogeneous row fieldType
instance homogeneous
  :: ( RowToList row fields
     , FieldOf fields fieldType
     )
  => Homogeneous (? row) fieldType -- not sure what to do here

class FieldOf (rowList :: RowList) fieldType
instance fieldOfCons
  :: (FieldOf tail fieldType)
  => FieldOf (Cons symbol fieldType tail) fieldType
instance fieldOfNil :: FieldOf Nil fieldType
```

Does this seems like I'm on the right track?

I'm not quite sure about the other 2 bullet points but eager to understand.

**Comment by LiamGoodacre** [comment]
@paulyoung 
The Homogeneous type class will need a kind annotation of `# Type` on the `row` parameter.
For the instance, all the arguments can be left as type variables.  I.e:
```purescript
instance homogeneous
  :: ( RowToList row fields
     , FieldOf fields fieldType
     )
  => Homogeneous row fieldType
```

Phil's point about `TypeEquals` is referring to the `FieldOf (Cons symbol fieldType tail) fieldType` instance.  The `fieldType` appears twice across multiple instance arguments.  This isn't necessarily bad, it's just that if there is a mismatch the error will mention that it couldn't find an instance for `FieldOf ...`.  We'd rather have an error that two types weren't equal.  So the action here is to rename one of the two usages of `fieldType` and add a new instance constraint requiring that those two types be equal.

The functional dependencies for `FieldOf` should be `| rowList -> fieldType`.  Because once we know which instance to select - which we do by looking at the `rowList`, i.e whether or not it is `Nil` or `Cons` - and the constraints have been satisfied, the `fieldType` should be fully determined.  The same sort of functional dependencies could be used on `Homogeneous`.  In cases where there is only one instance with all the arguments being type variables, I usually have both parameters always determined `| -> row fieldType` - but I think Phil sometimes disagrees with doing that? @paf31

**Comment by paf31** [comment]
I don't think we can have both parameters determined, since that says you can predict the row given no information.

**Comment by paulyoung** [comment]
Thanks for the helpful feedback so far. I pushed some changes.

**Comment by paulyoung** [comment]
Is there anything else I can do here?

File: src/Type/Row.purs [hunk 2]
```diff
@@ -70,3 +74,17 @@ instance listToRowCons
   :: ( ListToRow tail tailRow
      , RowCons label ty tailRow row )
   => ListToRow (Cons label ty tail) row
+
+
+class Homogeneous (row :: # Type) fieldType
```
Comment (paf31) [inline src/Type/Row.purs]: Also missing a functional dependency here.

Comment (paf31) [inline src/Type/Row.purs]: Could you please add a comment?

Comment (paf31) [inline src/Type/Row.purs]: Perhaps a value-level function `fieldType :: RProxy row -> Proxy fieldType` would be handy as well.

**Comment by paf31** [comment]
@LiamGoodacre Any thoughts on this? What do you think about creating some submodules here?

**Comment by LiamGoodacre** [comment]
Just been thinking about the fundeps and I'm thinking that this might suffer from the same issue I mentioned about multiple constraints.  I'll try to come up with a demonstration tonight.

**Comment by LiamGoodacre** [comment]
Actually, I am mistaken.  Ignore my previous concern.

**Comment by LiamGoodacre** [comment]
Submodules sound good.  What do you suggest here? @paf31

File: src/Type/Row.purs (hunk 2 above)
Comment (paulyoung) [reply src/Type/Row.purs]: Will do.

Comment (paulyoung) [reply src/Type/Row.purs]: Good catch.

Comment (paulyoung) [reply src/Type/Row.purs]: Am I correct in thinking that the right thing to do here is `| row -> fieldType`?

File: src/Type/Row.purs [hunk 3]
```diff
@@ -70,3 +74,17 @@ instance listToRowCons
   :: ( ListToRow tail tailRow
      , RowCons label ty tailRow row )
   => ListToRow (Cons label ty tail) row
+
+-- | Ensure that every field in a row has the same type.
+class Homogeneous (row :: # Type) fieldType
```
Comment (paulyoung) [inline src/Type/Row.purs]: I though I'd asked this already but can't see it now.

Is `| row -> fieldType` missing here?

Comment (paf31) [reply src/Type/Row.purs]: Yes I think so

**Comment by paulyoung** [comment]
Should I put this in a `Type.Row.Homogeneous` module?

**Comment by paf31** [comment]
> Should I put this in a Type.Row.Homogeneous module?

Sorry for the delay. Yes please.

File: src/Type/Row/Homogeneous.purs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+module Type.Row.Homogeneous
+  ( class FieldOf
+  , class Homogeneous
+  ) where
+
+-- | Ensure that every field in a row has the same type.
```
Comment (paf31) [inline src/Type/Row/Homogeneous.purs]: This comment should be removed, I think.

File: src/Type/Row/Homogeneous.purs [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import Type.Row (class RowToList, Cons, Nil, kind RowList)
+
+-- | Ensure that every field in a row has the same type.
+class Homogeneous (row :: # Type) fieldType | row -> fieldType
+instance homogeneous
+  :: ( RowToList row fields
+     , FieldOf fields fieldType )
+  => Homogeneous row fieldType
+
+class FieldOf (rowList :: RowList) fieldType | rowList -> fieldType
```
Comment (paf31) [inline src/Type/Row/Homogeneous.purs]: It's not clear to me (without reading the docs quite a bit) which one of these should be used in regular code. I'd suggest naming this one something like `HomogeneousRowList` to just be explicit about it.

File: src/Type/Row/Homogeneous.purs [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+instance fieldOfCons
+  :: ( FieldOf tail fieldType
+     , TypeEquals fieldType fieldType2 )
```
Comment (paf31) [inline src/Type/Row/Homogeneous.purs]: Does this need `TypeEquals`? Can you just use 

```purescript
instance fieldOfCons :: FieldOf tail fieldType => FieldOf (Cons symbol fieldType tail) fieldType
```

Comment (paulyoung) [reply src/Type/Row/Homogeneous.purs]: I did that in https://github.com/purescript/purescript-typelevel-prelude/pull/20/commits/8b39b00eab1e640a0fe3af7badcf3af0218240c3#diff-86491d4ab8759f2c9d61332f39afdd9bR15 but you suggested using `TypeEquals` in https://github.com/purescript/purescript-typelevel-prelude/pull/20#issuecomment-321603793 😄

File: src/Type/Row/Homogeneous.purs (hunk 4 above)
Comment (paulyoung) [reply src/Type/Row/Homogeneous.purs]: Yes. Strange things like this sometimes happen as a result of imports being managed by the PureScript Emacs layer.

File: src/Type/Row/Homogeneous.purs (hunk 6 above)
Comment (paf31) [reply src/Type/Row/Homogeneous.purs]: D'oh. Thanks :)

**Comment by paulyoung** [comment]
I think I've addressed everything in the latest round of feedback.

File: src/Type/Row/Homogeneous.purs [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+  :: ( RowToList row fields
+     , HomogenousRowList fields fieldType )
+  => Homogeneous row fieldType
+
+class HomogenousRowList (rowList :: RowList) fieldType | rowList -> fieldType
+instance fieldOfCons
+  :: ( HomogenousRowList tail fieldType
+     , TypeEquals fieldType fieldType2 )
+  => HomogenousRowList (Cons symbol fieldType tail) fieldType2
+instance fieldOfNil :: HomogenousRowList Nil fieldType
```
Comment (paulyoung) [inline src/Type/Row/Homogeneous.purs]: I suppose I should rename these to `homogenousRowListCons` and `homogenousRowListNil`

**Comment by paf31** [comment]
Thanks!

**Pull request merged**